csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+
+      new_out = ops::newValLike(new_pad_out, uop->out()->getDataType().value());
+      IrBuilder::create<UnaryOp>(uop->getUnaryOpType(), new_out, new_pad_out);
+      // insert new PadOp(s) to frontier;
+      frontier.push_back(new_pad_out->definition()->as<PadOp>());
+      simple_pad_set.insert(new_pad_out->definition()->as<PadOp>());
+    } else if (auto* bop = dynamic_cast<BinaryOp*>(def)) {
+      // stop propagation if current PadOp p isn't the only use of tv, since it
+      // requires tv to be live in the fusion.
+      if (tv->uses().size() != 1) {
+        continue;
+      }
+      NVF_ERROR(
+          tv->uses()[0] == p,
+          "expect existing PadOp to be the only use of its input");
+
+      // check if unary op type is compatible for zero pad propagation.
+      if (!zeroIsIdentity(bop->getBinaryOpType())) {
+        continue;
+      }
+      // check if PadOp can be replayed on input(s)
+      if (!pad_replay_check(bop)) {
+        continue;
+      }
+
+      // check for broadcast on padded axis.
+      auto* lhs = bop->lhs()->as<TensorView>();
+      auto* rhs = bop->rhs()->as<TensorView>();
+      bool pad_on_broadcast = false;
+      for (auto i : p->getPaddedAxes()) {
+        if (lhs->getLogicalDomain()[i]->isBroadcast() ||
+            rhs->getLogicalDomain()[i]->isBroadcast()) {
+          pad_on_broadcast = true;
+          break;
+        }
+      }
+      // padding on broadcast dimensions is not supported in propagation yet.
+      if (pad_on_broadcast) {
+        continue;
+      }
+
+      // replay pad on input(s)
+      std::vector<Val*> vals = {
+          replayConcretePad(
+              bop->lhs()->as<TensorView>(),
+              p->value(),
+              {p->getPadWidths()},
+              TensorDomain::noReductions(
+                  p->out()->as<TensorView>()->getLogicalDomain())),
+          replayConcretePad(
+              bop->rhs()->as<TensorView>(),
+              p->value(),
+              {p->getPadWidths()},
+              TensorDomain::noReductions(
+                  p->out()->as<TensorView>()->getLogicalDomain()))};
+
+      new_out = ops::newOutputTV(vals, bop->out()->getDataType().value());
+      IrBuilder::create<BinaryOp>(
+          bop->getBinaryOpType(), new_out, vals[0], vals[1]);
+      // insert new PadOp(s) to frontier;
+      frontier.push_back(vals[0]->definition()->as<PadOp>());
+      simple_pad_set.insert(vals[0]->definition()->as<PadOp>());
+      frontier.push_back(vals[1]->definition()->as<PadOp>());
+      simple_pad_set.insert(vals[1]->definition()->as<PadOp>());
+    } else if (auto* pop = dynamic_cast<PadOp*>(def)) {
+      // stop propagation if PadOp `pop` isn't a simple PadOp, since we can only
+      // merge simple PadOp together. Note that we don't need to check the other
+      // uses of `tv` here, since we want to merge the consecutive pads anyway
+      // and it won't interfere the other uses of `tv`.
+      if (simple_pad_set.count(pop) == 0) {
```

Comment:
It feels like we've maintained simple_pad_set just for the pad-pad corner case. Unless proven to be expensive, I'd call isSimplePadOp on demand, which is simpler.

Reply:
I think the issue is that while `isSimplePadOp` works as expected on the original pad ops it's not clear that it will still return true for the replacements we make along the way; mainly because we currently pattern match the uses for a `CatOp` which will fail once we shuffle the pad in the producer direction. So as the frontier evolves we need to track that the pads we encounter are either original simple pads or replacements from simple pads. This could use a comment in the code I think.

Reply:
Yeah Jacob is right. This is the same reason I have this hacky thing inside: https://github.com/NVIDIA/Fuser/blob/7557dfda306899b96e90bc98353cf47de7889619/csrc/preseg_passes/move_pad.cpp#L32-L37

I'll put a comment down where `simple_pad` is as well